diff -uNr zd1211-driver-r85.old/src/zdusb.c zd1211-driver-r85/src/zdusb.c
--- zd1211-driver-r85/src/zdusb.c	2007-02-25 00:31:21.000000000 +0100
+++ zd1211-driver-r85/src/zdusb.c	2008-05-26 21:06:29.000000000 +0200
@@ -85,6 +85,7 @@
                 { USB_DEVICE(0x0105, 0x145f) },
                 { USB_DEVICE(0x0586, 0x3401) },
                 { USB_DEVICE(0x0586, 0x3409) },
+                { USB_DEVICE(0x0586, 0x3410) },
                 { USB_DEVICE(0x0675, 0x0550) },
                 { USB_DEVICE(0x079b, 0x004a) },
                 { USB_DEVICE(0x07b8, 0x6001) },
